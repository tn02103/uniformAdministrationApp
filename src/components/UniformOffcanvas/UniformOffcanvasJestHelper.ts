import { Deficiency } from "@/types/deficiencyTypes";
import { generationLists, sizeLists, typeList } from "../../../tests/_jestConfig/staticMockData";
import { AuthRole } from "@/lib/AuthRoles";


// ------------- MOCKS FOR DEFICIENCY -------------
jest.mock('@/dal/inspection/deficiency', () => ({
    createUniformDeficiency: jest.fn(async () => "created successfully"),
    resolveDeficiency: jest.fn(async () => "resolved successfully"),
    updateUniformDeficiency: jest.fn(async () => "updated successfully"),
}));

jest.mock('@/dataFetcher/deficiency', () => ({
    useDeficienciesByUniformId: jest.fn((_, includeResolved) => {
        if (includeResolved) {
            return { deficiencies: mockDeficiencyList };
        } else {
            return { deficiencies: mockDeficiencyList.slice(0, 2) };
        }
    }),
    useDeficiencyTypes: jest.fn(() => ({ deficiencyTypeList: mockDeficiencyTypeList })),
}));

// ------------- MOCKS FOR UNIFORM -------------

jest.mock('@/dataFetcher/uniformAdmin', () => ({
    useUniformGenerationListByType: jest.fn(() => ({
        generationList: generationLists[0]
    })),
    useUniformTypeList: jest.fn(() => ({
        typeList
    })),
}));
jest.mock("@/dal/uniform/item/_index", () => ({
    updateUniformItem: jest.fn(() => Promise.resolve('Saved item')),
    deleteUniformItem: jest.fn(() => Promise.resolve('Deleted item')),
    getUniformItemHistory: jest.fn(() => Promise.resolve([])), // Mock for history
    getUniformItemCountByType: jest.fn(() => Promise.resolve(0)), // Mock for count
    issueUniformItem: jest.fn(),
    createUniformItems: jest.fn(),
    getUniformItemDeficiencies: jest.fn(),
}));
jest.mock('../globalDataProvider', () => ({
    useGlobalData: jest.fn(() => ({
        sizelists: sizeLists,
        userRole: global.__ROLE__ ?? AuthRole.admin,
    })),
}));
jest.mock('@/dataFetcher/uniform', () => ({
    useUniformItemHistory: jest.fn(() => ({
        history: mockUniformHistory,
    })),
}));

// ------------- OTHER MOCKS -------------
jest.mock("swr", () => ({
    mutate: jest.fn(async () => { }),
}));

jest.mock('react-toastify', () => ({
    toast: {
        error: jest.fn(),
        success: jest.fn(),
    },
}));

export const mockDeficiencyTypeList = [
    {
        id: 'de3860d4-c88e-4a7c-be4c-7e832eda31d4',
        name: 'Test Type',
        dependent: 'uniform',
    },
    {
        id: '61d28738-4757-4e98-bfbf-a82f8404a74a',
        name: 'Another One',
        dependent: 'uniform',
    },
];
export const mockDeficiencyList: Deficiency[] = [
    {
        id: '80c36a66-81e9-4c9c-8d56-a095a6e9e28b',
        comment: 'Test comment',
        description: 'Test description',
        dateCreated: new Date('2023-10-01T12:00:00Z'),
        dateUpdated: new Date('2023-10-01T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user',
        typeId: mockDeficiencyTypeList[0].id,
        typeName: mockDeficiencyTypeList[0].name,
    },
    {
        id: 'a1b2c3d4-e5f6-7g8h-9i0j-k1l2m3n4o5p6',
        comment: 'Test comment 2',
        description: 'Test description 2',
        dateCreated: new Date('2023-10-02T12:00:00Z'),
        dateUpdated: new Date('2023-10-12T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user1',
        typeId: mockDeficiencyTypeList[1].id,
        typeName: mockDeficiencyTypeList[1].name,
    },
    {
        id: 'b7c8d9e0-f1g2-h3i4-j5k6-l7m8n9o0p1q2',
        comment: 'Test comment 3',
        description: 'Test description 3',
        dateCreated: new Date('2023-10-03T12:00:00Z'),
        dateUpdated: new Date('2023-10-12T12:00:00Z'),
        dateResolved: new Date('2023-10-17T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user1',
        userResolved: 'user2',
        typeId: mockDeficiencyTypeList[0].id,
        typeName: mockDeficiencyTypeList[0].name,
    },
];
export const mockUniform = {
    id: "c227ac23-93d4-42b5-be2e-956ea35c2db9",
    number: 2501,
    generation: generationLists[0][1],
    size: sizeLists[0].uniformSizes[0],
    comment: "Test comment",
    active: true,
    type: {
        id: typeList[0].id,
        name: typeList[0].name,
    },
};
export const mockUniformHistory = [
    {
        dateIssued: new Date('2023-10-01T12:00:00Z'),
        dateReturned: null,
        cadet: {
            firstname: 'John',
            lastname: 'Doe',
            recdelete: false,
        },
    },
    {
        dateIssued: new Date('2023-10-02T12:00:00Z'),
        dateReturned: new Date('2023-10-03T12:00:00Z'),
        cadet: {
            firstname: 'Jane',
            lastname: 'Smith',
            recdelete: true,
        },
    },
];