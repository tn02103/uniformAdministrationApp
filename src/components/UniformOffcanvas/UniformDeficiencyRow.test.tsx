import "./UniformOffcanvasJestHelper";

import { getByLabelText, getByRole, getByText, queryByRole, queryByText, render, screen } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import { UniformDeficiencyRow } from "./UniformDeficiencyRow";
import { mockDeficiencyList, mockDeficiencyTypeList, mockUniform } from "./UniformOffcanvasJestHelper";

describe('UniformDeficiencyRow', () => {
    const setEditable = async (card: HTMLElement, user: UserEvent) => {
        const actionMenu = getByRole(card, 'button', { name: /deficiency.label.actions/i });
        await user.click(actionMenu);
        const editButton = getByRole(card, 'button', { name: /edit/i });
        await user.click(editButton);
    }

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { container } = render(
            <UniformDeficiencyRow
                uniformId={mockUniform.id}
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('only shows resolved deficiencies when includeResolved is true', async () => {
        const user = userEvent.setup();
        const { useDeficienciesByUniformId } = jest.requireMock('@/dataFetcher/deficiency');

        render(
            <UniformDeficiencyRow
                uniformId={mockUniform.id}
            />
        );

        const showResolvedToggle = screen.getByRole('checkbox', { name: /includeResolved/i });
        expect(showResolvedToggle).not.toBeChecked();
        expect(useDeficienciesByUniformId).toHaveBeenCalledTimes(1);
        expect(useDeficienciesByUniformId).toHaveBeenCalledWith(mockUniform.id, false);
        expect(screen.getAllByRole('listitem')).toHaveLength(2);

        await user.click(showResolvedToggle);
        expect(showResolvedToggle).toBeChecked();
        expect(useDeficienciesByUniformId).toHaveBeenCalledTimes(2);
        expect(useDeficienciesByUniformId).toHaveBeenLastCalledWith(mockUniform.id, true);
        expect(screen.getAllByRole('listitem')).toHaveLength(3);

        await user.click(showResolvedToggle);
        expect(showResolvedToggle).not.toBeChecked();
        expect(useDeficienciesByUniformId).toHaveBeenCalledTimes(3);
        expect(useDeficienciesByUniformId).toHaveBeenLastCalledWith(mockUniform.id, false);
        expect(screen.getAllByRole('listitem')).toHaveLength(2);
    });

    it('shows empty state when no deficiencies', () => {
        const { useDeficienciesByUniformId } = jest.requireMock('@/dataFetcher/deficiency');
        useDeficienciesByUniformId.mockReturnValueOnce({ deficiencies: [] });

        render(
            <UniformDeficiencyRow
                uniformId={mockUniform.id}
            />
        );

        expect(screen.getByText(/noDeficiencies/i)).toBeInTheDocument();
    });

    it('doesnt show actionMenu when resolved', async () => {
        const user = userEvent.setup();
        render(
            <UniformDeficiencyRow
                uniformId={mockUniform.id}
            />
        );

        const showResolvedToggle = screen.getByRole('checkbox', { name: /includeResolved/i });
        await user.click(showResolvedToggle);

        const card = screen.getAllByRole('listitem')[2];
        expect(card).toBeInTheDocument();
        const actionMenu = queryByRole(card, 'button', { name: /deficiency.label.actions/i });
        expect(actionMenu).toBeNull();
    });

    describe('expand/collapse', () => {
        it('should expand/collapse the card - unresolved', async () => {
            const user = userEvent.setup();
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            const card = screen.getAllByRole('listitem')[1];
            expect(card).toBeInTheDocument();

            const expandButton = getByRole(card, 'button', { name: /expandableArea.showMore/i });
            await user.click(expandButton);

            expect(getByText(card, /label.date.created/i)).toBeInTheDocument();
            expect(getByText(card, /label.date.updated/i)).toBeInTheDocument();
            expect(queryByText(card, /label.date.deleted/i)).toBeNull();
            expect(getByText(card, /label.user.created/i)).toBeInTheDocument();
            expect(getByText(card, /label.user.updated/i)).toBeInTheDocument();
            expect(queryByText(card, /label.user.deleted/i)).toBeNull();

            const dateCreated = getByLabelText(card, /label.date.created/i);
            const dateUpdated = getByLabelText(card, /label.date.updated/i);
            const userCreated = getByLabelText(card, /label.user.created/i);
            const userUpdated = getByLabelText(card, /label.user.updated/i);

            expect(dateCreated).toHaveTextContent('02.10.2023');
            expect(dateUpdated).toHaveTextContent('12.10.2023');
            expect(userCreated).toHaveTextContent('user');
            expect(userUpdated).toHaveTextContent('user1');

            const collapseButton = getByRole(card, 'button', { name: /showLess/i });
            await user.click(collapseButton);

            expect(queryByText(card, /label.date.created/i)).toBeNull();
            expect(queryByText(card, /label.date.updated/i)).toBeNull();
            expect(queryByText(card, /label.user.created/i)).toBeNull();
            expect(queryByText(card, /label.user.updated/i)).toBeNull();
        });

        it('should expand/collapse the card - resolved', async () => {
            const user = userEvent.setup();
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );
            const showResolvedToggle = screen.getByRole('checkbox', { name: /includeResolved/i });
            await user.click(showResolvedToggle);
            expect(showResolvedToggle).toBeChecked();

            const card = screen.getAllByRole('listitem')[2];
            expect(card).toBeInTheDocument();

            const expandButton = getByRole(card, 'button', { name: /expandableArea.showMore/i });
            await user.click(expandButton);

            expect(getByText(card, /label.date.created/i)).toBeInTheDocument();
            expect(getByText(card, /label.date.updated/i)).toBeInTheDocument();
            expect(getByText(card, /label.date.resolved/i)).toBeInTheDocument();
            expect(getByText(card, /label.user.created/i)).toBeInTheDocument();
            expect(getByText(card, /label.user.updated/i)).toBeInTheDocument();
            expect(getByText(card, /label.user.resolved/i)).toBeInTheDocument();

            const dateCreated = getByLabelText(card, /label.date.created/i);
            const dateUpdated = getByLabelText(card, /label.date.updated/i);
            const dateDeleted = getByLabelText(card, /label.date.resolved/i);
            const userCreated = getByLabelText(card, /label.user.created/i);
            const userUpdated = getByLabelText(card, /label.user.updated/i);
            const userDeleted = getByLabelText(card, /label.user.resolved/i);

            expect(dateCreated).toHaveTextContent('03.10.2023');
            expect(dateUpdated).toHaveTextContent('12.10.2023');
            expect(dateDeleted).toHaveTextContent('17.10.2023');
            expect(userCreated).toHaveTextContent('user');
            expect(userUpdated).toHaveTextContent('user1');
            expect(userDeleted).toHaveTextContent('user2');

            const collapseButton = getByRole(card, 'button', { name: /showLess/i });
            await user.click(collapseButton);

            expect(queryByText(card, /label.date.created/i)).toBeNull();
            expect(queryByText(card, /label.date.updated/i)).toBeNull();
            expect(queryByText(card, /label.date.resolved/i)).toBeNull();
            expect(queryByText(card, /label.user.created/i)).toBeNull();
            expect(queryByText(card, /label.user.updated/i)).toBeNull();
            expect(queryByText(card, /label.user.resolved/i)).toBeNull();
        });
    });

    describe('createDeficiencyCard', () => {
        it('should show createDeficiencyCard', async () => {
            const user = userEvent.setup();
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            await user.click(screen.getByRole('button', { name: /create/i }));
            const createCard = screen.getByRole('listitem', { name: /createCardLabel/i });

            const createButton = getByRole(createCard, 'button', { name: /create/i });
            const cancelButton = getByRole(createCard, 'button', { name: /cancel/i });
            const commentInput = getByRole(createCard, 'textbox', { name: /comment/i });
            const typeSelect = getByRole(createCard, 'combobox', { name: /deficiencyType/i });

            expect(createButton).toBeEnabled();
            expect(cancelButton).toBeEnabled();
            expect(commentInput).toHaveValue('');
            expect(typeSelect).toHaveValue(mockDeficiencyTypeList[0].id);

            expect(typeSelect).toHaveTextContent(mockDeficiencyTypeList[0].name);
            expect(getByRole(typeSelect, 'option', { name: mockDeficiencyTypeList[0].name })).toBeInTheDocument();
            expect(getByRole(typeSelect, 'option', { name: mockDeficiencyTypeList[1].name })).toBeInTheDocument();
        });

        it('should hide on cancel', async () => {
            const user = userEvent.setup();
            const { container } = render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            await user.click(getByRole(container, 'button', { name: /create/i }));
            const createCard = getByRole(container, 'listitem', { name: /createCardLabel/i });
            expect(createCard).toBeInTheDocument();

            const cancelButton = getByRole(createCard, 'button', { name: /cancel/i });
            await user.click(cancelButton);
            expect(createCard).not.toBeInTheDocument();
        });

        it('should call createDeficiency on create', async () => {
            const user = userEvent.setup();
            const { createUniformDeficiency } = jest.requireMock('@/dal/inspection/deficiency');
            const { mutate } = jest.requireMock('swr');
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            await user.click(screen.getByRole('button', { name: /create/i }));
            const createCard = screen.getByRole('listitem', { name: /createCardLabel/i });
            const createButton = getByRole(createCard, 'button', { name: /create/i });
            const commentInput = getByRole(createCard, 'textbox', { name: /comment/i });
            const typeSelect = getByRole(createCard, 'combobox', { name: /deficiencyType/i });

            await user.type(commentInput, 'Test comment');
            await user.selectOptions(typeSelect, mockDeficiencyTypeList[1].id);
            expect(createButton).toBeEnabled();

            await user.click(createButton);
            expect(createUniformDeficiency).toHaveBeenCalledTimes(1);
            expect(createUniformDeficiency).toHaveBeenCalledWith({
                uniformId: mockUniform.id,
                data: {
                    comment: 'Test comment',
                    typeId: mockDeficiencyTypeList[1].id
                }
            });
            expect(mutate).toHaveBeenCalledTimes(1);
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.deficiencies.true`)).toBeTruthy();
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.deficiencies.false`)).toBeTruthy();
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.somethingElse`)).toBeFalsy();

            expect(screen.queryByRole('listitem', { name: /createCardLabel/i })).not.toBeInTheDocument();
        });

        it('should catch exceptions on create', async () => {
            const user = userEvent.setup();
            const { createUniformDeficiency } = jest.requireMock('@/dal/inspection/deficiency');
            const { mutate } = jest.requireMock('swr');
            const { toast } = jest.requireMock('react-toastify');
            createUniformDeficiency.mockRejectedValueOnce(new Error('Test error'));

            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            await user.click(screen.getByRole('button', { name: /create/i }));
            const createCard = screen.getByRole('listitem', { name: /createCardLabel/i });

            const createButton = getByRole(createCard, 'button', { name: /create/i });
            const commentInput = getByRole(createCard, 'textbox', { name: /comment/i });
            const typeSelect = getByRole(createCard, 'combobox', { name: /deficiencyType/i });

            await user.type(commentInput, 'Test comment');
            await user.selectOptions(typeSelect, mockDeficiencyTypeList[1].id);
            await user.click(createButton);

            expect(createUniformDeficiency).toHaveBeenCalledTimes(1);
            expect(mutate).toHaveBeenCalledTimes(0);
            expect(toast.error).toHaveBeenCalledTimes(1);
        });
    });

    describe('update Deficiency', () => {
        it('sets card to edit mode', async () => {
            const user = userEvent.setup();
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );
            // Set the card to edit mode
            const firstCard = screen.getAllByRole('listitem')[0];
            await setEditable(firstCard, user);

            // get elements
            const saveButton = getByRole(firstCard, 'button', { name: /save/i });
            const cancelButton = getByRole(firstCard, 'button', { name: /cancel/i });
            const commentInput = getByRole(firstCard, 'textbox', { name: /comment/i });
            const typeSelect = getByRole(firstCard, 'combobox', { name: /deficiencyType/i });

            // check that the card is in edit mode
            expect(saveButton).toBeEnabled();
            expect(cancelButton).toBeEnabled();
            expect(commentInput).toHaveValue(mockDeficiencyList[0].comment);
            expect(typeSelect).toHaveValue(mockDeficiencyList[0].typeId);

            // validate typeSelect options
            expect(typeSelect).toHaveTextContent(mockDeficiencyTypeList[0].name);
            expect(getByRole(typeSelect, 'option', { name: mockDeficiencyTypeList[0].name })).toBeInTheDocument();
            expect(getByRole(typeSelect, 'option', { name: mockDeficiencyTypeList[1].name })).toBeInTheDocument();
        });

        it('resets data on cancel', async () => {
            const user = userEvent.setup();
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            // Set the card to edit mode
            const firstCard = screen.getAllByRole('listitem')[0];
            await setEditable(firstCard, user);

            // change form data
            const commentInput = getByRole(firstCard, 'textbox', { name: /comment/i });
            const typeSelect = getByRole(firstCard, 'combobox', { name: /deficiencyType/i });
            await user.clear(commentInput);
            await user.type(commentInput, 'Something creative');
            await user.selectOptions(typeSelect, mockDeficiencyTypeList[1].id);

            // cancel the edit
            const cancelButton = getByRole(firstCard, 'button', { name: /cancel/i });
            await user.click(cancelButton);

            // check that the card is not in edit mode
            expect(typeSelect).not.toBeInTheDocument();
            expect(commentInput).not.toBeInTheDocument();

            // check that the data is reset
            expect(getByText(firstCard, mockDeficiencyList[0].comment)).toBeInTheDocument();
            expect(getByText(firstCard, mockDeficiencyList[0].typeName)).toBeInTheDocument();
            expect(queryByText(firstCard, 'Something creative')).toBeNull();
            expect(queryByText(firstCard, mockDeficiencyTypeList[1].name)).toBeNull();

            // check that formValues are reset when editing again
            await setEditable(firstCard, user);
            expect(getByRole(firstCard, 'textbox', { name: /comment/i })).toHaveValue(mockDeficiencyList[0].comment);
            expect(getByRole(firstCard, 'combobox', { name: /deficiencyType/i })).toHaveValue(mockDeficiencyList[0].typeId);
        });
        it('should call updateDeficiency on save', async () => {
            const user = userEvent.setup();
            const { updateUniformDeficiency } = jest.requireMock('@/dal/inspection/deficiency');
            const { mutate } = jest.requireMock('swr');
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            // Set the card to edit mode
            const firstCard = screen.getAllByRole('listitem')[0];
            await setEditable(firstCard, user);

            // change form data
            const commentInput = getByRole(firstCard, 'textbox', { name: /comment/i });
            const typeSelect = getByRole(firstCard, 'combobox', { name: /deficiencyType/i });
            await user.clear(commentInput);
            await user.type(commentInput, 'Something creative');
            await user.selectOptions(typeSelect, mockDeficiencyTypeList[1].id);

            // save the edit
            const saveButton = getByRole(firstCard, 'button', { name: /save/i });
            await user.click(saveButton);

            // validate function calls
            expect(updateUniformDeficiency).toHaveBeenCalledTimes(1);
            expect(updateUniformDeficiency).toHaveBeenCalledWith({
                id: mockDeficiencyList[0].id,
                data: {
                    comment: 'Something creative',
                    typeId: mockDeficiencyTypeList[1].id
                }
            });
            expect(mutate).toHaveBeenCalledTimes(1);
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.deficiencies.true`)).toBeTruthy();
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.deficiencies.false`)).toBeTruthy();
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.somethingElse`)).toBeFalsy();

            // check that the card is not in edit mode
            expect(commentInput).not.toBeInTheDocument();
            expect(typeSelect).not.toBeInTheDocument();
        });
        it('should catch exceptions on update', async () => {
            const user = userEvent.setup();
            const { updateUniformDeficiency } = jest.requireMock('@/dal/inspection/deficiency');
            const { mutate } = jest.requireMock('swr');
            const { toast } = jest.requireMock('react-toastify');
            updateUniformDeficiency.mockRejectedValueOnce(new Error('Test error'));

            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            // Set the card to edit mode
            const firstCard = screen.getAllByRole('listitem')[0];
            await setEditable(firstCard, user);

            // save the edit
            const saveButton = getByRole(firstCard, 'button', { name: /save/i });
            await user.click(saveButton);

            // validate function calls
            expect(updateUniformDeficiency).toHaveBeenCalledTimes(1);
            expect(mutate).toHaveBeenCalledTimes(0);
            expect(toast.error).toHaveBeenCalledTimes(1);

            // check that the card is still in edit mode
            expect(getByRole(firstCard, 'textbox', { name: /comment/i })).toBeInTheDocument();
            expect(getByRole(firstCard, 'combobox', { name: /deficiencyType/i })).toBeInTheDocument();
            expect(getByRole(firstCard, 'button', { name: /save/i })).toBeInTheDocument();
        });
    });
    describe('resolve Deficiency', () => {
        it('calls resolveDeficiency', async () => {
            const user = userEvent.setup();
            const { resolveDeficiency } = jest.requireMock('@/dal/inspection/deficiency');
            const { mutate } = jest.requireMock('swr');
            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            // resolve the deficiency
            const firstCard = screen.getAllByRole('listitem')[0];
            const actionMenu = getByRole(firstCard, 'button', { name: /deficiency.label.actions/i });
            await user.click(actionMenu);
            const resolveButton = getByRole(firstCard, 'button', { name: /resolve/i });
            await user.click(resolveButton);

            // validate function calls
            expect(resolveDeficiency).toHaveBeenCalledTimes(1);
            expect(resolveDeficiency).toHaveBeenCalledWith(mockDeficiencyList[0].id);
            expect(mutate).toHaveBeenCalledTimes(1);
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.deficiencies.true`)).toBeTruthy();
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.deficiencies.false`)).toBeTruthy();
            expect(mutate.mock.calls[0][0](`uniform.${mockUniform.id}.somethingElse`)).toBeFalsy();
        });
        it('should catch exceptions on resolve', async () => {
            const user = userEvent.setup();
            const { resolveDeficiency } = jest.requireMock('@/dal/inspection/deficiency');
            const { mutate } = jest.requireMock('swr');
            const { toast } = jest.requireMock('react-toastify');
            resolveDeficiency.mockRejectedValueOnce(new Error('Test error'));

            render(
                <UniformDeficiencyRow
                    uniformId={mockUniform.id}
                />
            );

            // resolve the deficiency
            const firstCard = screen.getAllByRole('listitem')[0];
            const actionMenu = getByRole(firstCard, 'button', { name: /deficiency.label.actions/i });
            await user.click(actionMenu);
            const resolveButton = getByRole(firstCard, 'button', { name: /resolve/i });
            await user.click(resolveButton);

            // validate error handling
            expect(resolveDeficiency).toHaveBeenCalledTimes(1);
            expect(mutate).toHaveBeenCalledTimes(0);
            expect(toast.error).toHaveBeenCalledTimes(1);
        });
    });
});