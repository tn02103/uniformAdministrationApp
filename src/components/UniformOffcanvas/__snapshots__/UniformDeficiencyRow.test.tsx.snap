// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniformDeficiencyRow renders correctly 1`] = `
<div>
  <div
    class="m-0 mb-2 justify-content-between row"
  >
    <div
      class="col-sm-auto col-5"
    >
      <button
        class="mb-2 border-0 btn btn-outline-success"
        type="button"
      >
        common.actions.create
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plus fa-sm "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="text-start m-2 col-sm-auto col-6"
    >
      <div
        class="form-check form-switch"
      >
        <input
          aria-label="uniformOffcanvas.deficiency.includeResolved"
          class="form-check-input"
          name="showResolved"
          type="checkbox"
        />
        <label
          class="form-check-label"
          title=""
        >
          uniformOffcanvas.deficiency.includeResolved
        </label>
      </div>
    </div>
  </div>
  <div
    aria-label="Deficiency list"
    class="m-0 row"
    role="list"
  >
    <div
      aria-label="uniformOffcanvas.deficiency.cardLabel"
      class="m-1 p-0  card"
      role="listitem"
    >
      <div
        class="position-relative card-body"
      >
        <form
          autocomplete="off"
          class="mb-4"
          novalidate=""
        >
          <div
            aria-label="uniformOffcanvas.deficiency.label.deficiencyType"
            class="fs-6 fw-bold card-title h5"
          >
            Test Type
             
          </div>
          <div
            class="position-absolute top-0 end-0"
          >
            <div
              class="dropstart"
            >
              <button
                aria-expanded="false"
                aria-label="uniformOffcanvas.deficiency.label.actions"
                class="border-0 dropdown-toggle btn btn-outline-secondary"
                id="Cadetdropdown"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-ellipsis-vertical "
                  data-icon="ellipsis-vertical"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 128 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
          <p
            aria-label="uniformOffcanvas.deficiency.label.comment"
            class="card-text"
          >
            Test comment
          </p>
          <div
            class="align-items-center justify-content-center row"
            style="height: 2rem;"
          >
            <div
              class="position-relative"
            >
              <hr
                class=""
              />
              <div
                class="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
              >
                <button
                  class="border-0 btn btn-light btn-sm"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-chevron-down text-dark me-2 false"
                    data-icon="chevron-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                      fill="currentColor"
                    />
                  </svg>
                  expandableArea.showMore
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      aria-label="uniformOffcanvas.deficiency.cardLabel"
      class="m-1 p-0  card"
      role="listitem"
    >
      <div
        class="position-relative card-body"
      >
        <form
          autocomplete="off"
          class="mb-4"
          novalidate=""
        >
          <div
            aria-label="uniformOffcanvas.deficiency.label.deficiencyType"
            class="fs-6 fw-bold card-title h5"
          >
            Another One
             
          </div>
          <div
            class="position-absolute top-0 end-0"
          >
            <div
              class="dropstart"
            >
              <button
                aria-expanded="false"
                aria-label="uniformOffcanvas.deficiency.label.actions"
                class="border-0 dropdown-toggle btn btn-outline-secondary"
                id="Cadetdropdown"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-ellipsis-vertical "
                  data-icon="ellipsis-vertical"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 128 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
          <p
            aria-label="uniformOffcanvas.deficiency.label.comment"
            class="card-text"
          >
            Test comment 2
          </p>
          <div
            class="align-items-center justify-content-center row"
            style="height: 2rem;"
          >
            <div
              class="position-relative"
            >
              <hr
                class=""
              />
              <div
                class="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
              >
                <button
                  class="border-0 btn btn-light btn-sm"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-chevron-down text-dark me-2 false"
                    data-icon="chevron-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                      fill="currentColor"
                    />
                  </svg>
                  expandableArea.showMore
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
